### PR TITLE
Implement deprovision methods

### PR DESCRIPTION
**Pull request by MKodde**
## Why?
In order to further automate the GDPR rules, it's beneficial to connect StepUp to the OpenConext UserLifecycle application. This automates the removal of inactive users from the platform. And allows for a practical way to retrieve all user data from the StepUp platform. And also allows for exercising the right to be forgotten.

## How?
To achieve this goal, I added a deprovision/userlifecycle endpoint to the Middleware API.

This API exposes two endpoints. One to retrieve data, and one to remove user data.

### Retrieving data
Data retrieval is done in a very practical manner. The events that are stored for the identiy that needs/wants to remove its data is collected from the event store. This data contains all relevant data for the user. Including sensitive data. That data is then serialized to JSON data and returned to the lifecycle app.

Todo: the data is now aggregated in the IdentityRepository. This is no longer the case when @VadimSchmitz work is finished on task 3 of [#181430595](https://www.pivotaltracker.com/story/show/181430595).

### Removing data
Removing user data from StepUp was already implemented into the application. One hurdle remained, that was to expose this command to the LifeCycle API.

## More info
https://www.pivotaltracker.com/story/show/181430595

### CODE REVIEW

File: src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php [hunk 1]
```diff
@@ -106,4 +108,37 @@ public function test_it_can_return_data()
...
+        $this->apiRepo
+            ->shouldReceive('findOneByNameId')
+            ->with('urn:collab:person:example.com:maynard_keenan')
+            ->once()
+            ->andReturnNull();
+        $this->pipeline
+            ->shouldNotHaveReceived('process');
+        $this->deprovisionService->deprovision('urn:collab:person:example.com:maynard_keenan');
+    }
+    public function test_deprovision_method_performs_the_right_to_be_forgotten_command()
```
Comment (VadimSchmitz) [inline src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php:123]: The test does not perform any assertions, this leads to an incomplete test.
Is there a reason for this?

File: src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+        $this->eventRepo->shouldReceive('obtainInformation')->andReturn(['status' => 'OK', 'data' => []]);
+
+        $data = $this->deprovisionService->readUserData('urn:collab:person:example.com:maynard_keenan');
+
+        $this->assertTrue(is_array($data));
+        $this->assertEquals($data['status'], 'OK');
+    }
+
+    public function test_deprovision_does_not_deprovision_when_user_is_not_found()
```
Comment (VadimSchmitz) [inline src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php:112]: The test does not perform any assertions, this leads to an incomplete test.
Is there a reason for this?

File: src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/StepupMiddleware/ApiBundle/Tests/Service/DeprovisionServiceTest.php:120]: The test `test_deprovision_does_not_deprovision_when_user_is_not_found` indeed does not have a traditional PHPUnit assertion. Yielding the test warning. 

I thought I remedied this by adding the m::close() method call in the teardown method. This is not complete. I updated this by using the `MockeryPHPUnitIntegration` trait.

File: src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php [hunk 3]
```diff
@@ -18,32 +18,76 @@
...
+
+    public function obtainInformation(IdentityId $id): array
+    {
+        try {
+            $domainEventStream = $this->events->load($id);
+            $data = [];
+            /** @var DomainMessage $domainMessage */
+            foreach ($domainEventStream as $domainMessage) {
+                $event = $domainMessage->getPayload();
+                if (!$event instanceof Serializable) {
```
Comment (pablothedude) [inline src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php:70]: Should this also be checked also for the sensitive data?

File: src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php [hunk 4]
```diff
@@ -18,32 +18,76 @@
...
+        LoggerInterface $logger,
         array $eventStreamDecorators = []
     ) {
+        $this->events = $eventStore;
+        $this->logger = $logger;
         parent::__construct(
             $eventStore,
             $eventBus,
-            \Surfnet\Stepup\Identity\Identity::class,
+            Identity::class,
```
Comment (pablothedude) [inline src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php:56]: :100:

File: src/Surfnet/StepupMiddleware/ApiBundle/Controller/DeprovisionController.php [hunk 5]
```diff
@@ -18,20 +18,46 @@
...
+    public function deprovisionAction(string $collabPersonId): JsonResponse
     {
         $this->denyAccessUnlessGranted(['ROLE_DEPROVISION']);
-        return new JsonResponse();
+
+        $userData = $this->deprovisionService->readUserData($collabPersonId);
+        if (!empty($userData)) {
+            $this->deprovisionService->deprovision($collabPersonId);
+        }
+        return new JsonResponse($this->formatResponse('OK', $userData));
```
Comment (pablothedude) [inline src/Surfnet/StepupMiddleware/ApiBundle/Controller/DeprovisionController.php:45]: Should this return OK when a user is not found by the DeprovisioningService?

**Review by pablothedude — APPROVED** [review]

File: src/Surfnet/StepupMiddleware/ApiBundle/Service/DeprovisionService.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                )
+            );
+            return [];
+        }
+    }
+
+    public function deprovision(string $collabPersonId): void
+    {
+        try {
+            $user = $this->getIdentityByNameId($collabPersonId);
```
Comment (pablothedude) [inline src/Surfnet/StepupMiddleware/ApiBundle/Service/DeprovisionService.php:84]: Maybe you want to log this?

File: src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php (hunk 3 above)
Comment (MKodde) [reply src/Surfnet/Stepup/Identity/EventSourcing/IdentityRepository.php:70]: In parallel, @VadimSchmitz is working on a means to retrieve the user data directly from the Events. A next PR will integrate those changes here. Removing this 'pseudo' code that is currently in place.

File: src/Surfnet/StepupMiddleware/ApiBundle/Controller/DeprovisionController.php (hunk 5 above)
Comment (MKodde) [reply src/Surfnet/StepupMiddleware/ApiBundle/Controller/DeprovisionController.php:45]: Yes, good point! I will ensure that other status responses will not return a 200, OK response.

For now I'm leaving this as it is.

File: src/Surfnet/StepupMiddleware/ApiBundle/Service/DeprovisionService.php (hunk 6 above)
Comment (MKodde) [reply src/Surfnet/StepupMiddleware/ApiBundle/Service/DeprovisionService.php:84]: Good idea, I added some debug logging

**Review by VadimSchmitz — APPROVED** [review]

**Comment by MKodde** [comment]
Thanks for telling me IRL that the tests now fail on a sec warning. We will fix that issue in a seperate PR!

**Pull request merged**